internal/health_checks/api_check.go

```diff
@@ -0,0 +1,167 @@
+	metricType := "agent.googleapis.com/agent/ops_agent/enabled_receivers"
+	now := &timestamppb.Timestamp{
+		Seconds: time.Now().Unix(),
+	}
+	value := &monitoringpb.TypedValue{
+		Value: &monitoringpb.TypedValue_Int64Value{
+			Int64Value: int64(0),
+		},
+	}
+	req := &monitoringpb.CreateTimeSeriesRequest{
+		Name: "projects/" + gceMetadata.Project,
+		TimeSeries: []*monitoringpb.TimeSeries{{
+			MetricKind: metricpb.MetricDescriptor_GAUGE,
+			ValueType:  metricpb.MetricDescriptor_INT64,
+			Metric: &metricpb.Metric{
+				Type: metricType,
+			},
+			Resource: &monitoredres.MonitoredResource{
+				Type: "gce_instance",
+				Labels: map[string]string{
+					"instance_id": gceMetadata.InstanceID,
+					"zone":        gceMetadata.Zone,
+				},
+			},
+			Points: []*monitoringpb.Point{{
+				Interval: &monitoringpb.TimeInterval{
+					StartTime: now,
+					EndTime:   now,
+				},
+				Value: value,
+			}},
+		}},
+	}
+
+	return client.CreateTimeSeries(ctx, req)
+}
+
+type APICheck struct{}
+
+func (c APICheck) Name() string {
+	return "API Check"
+}
+
+func (c APICheck) RunCheck(logger *log.Logger) error {
+	ctx := context.Background()
+	gceMetadata, err := getGCEMetadata()
+	if err != nil {
+		return fmt.Errorf("can't get GCE metadata: %w", err)
+	}
+	logger.Println(gceMetadata)
+
+	// New Logging Client
+	logClient, err := logging.NewClient(ctx, gceMetadata.Project)
+	if err != nil {
+		return err
+	}
+	logger.Printf("logging client was created successfully.")
+
+	if err := logClient.Ping(ctx); err != nil {
+		logger.Println(err)
+		var apiErr *apierror.APIError
+		if errors.As(err, &apiErr) {
+			switch apiErr.Reason() {
+			case "SERVICE_DISABLED":
+				return LOG_API_DISABLED_ERR
+			case "ACCESS_TOKEN_SCOPE_INSUFFICIENT":
+				return LOG_API_SCOPE_ERR
+			case "IAM_PERMISSION_DENIED":
+				return LOG_API_PERMISSION_ERR
+			}
+
+			switch apiErr.GRPCStatus().Code() {
+			case codes.PermissionDenied:
+				return LOG_API_PERMISSION_ERR
+			case codes.Unauthenticated:
+				return LOG_API_SCOPE_ERR
+			}
+		}
+
+		return err
+	}
+	logClient.Close()
+
+	// New Monitoring Client
+	monClient, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		return err
+	}
+	logger.Printf("monitoring client was created successfully.")
+
+	if err := monitoringPing(ctx, *monClient, gceMetadata); err != nil {
+		logger.Println(err)
+		var apiErr *apierror.APIError
+		if errors.As(err, &apiErr) {
+			switch apiErr.Reason() {
+			case "SERVICE_DISABLED":
+				return MON_API_DISABLED_ERR
+			case "ACCESS_TOKEN_SCOPE_INSUFFICIENT":
+				return MON_API_SCOPE_ERR
+			case "IAM_PERMISSION_DENIED":
+				return MON_API_PERMISSION_ERR
+			}
+
+			switch apiErr.GRPCStatus().Code() {
+			case codes.PermissionDenied:
+				return MON_API_PERMISSION_ERR
+			case codes.Unauthenticated:
+				return MON_API_SCOPE_ERR
+			}
+		}
+		return err
+	}
+	monClient.Close()
```

Comment:
Also defer this.

Reply:
Addressed in https://github.com/GoogleCloudPlatform/ops-agent/pull/1051/commits/090592094044eaa06e4677874f20db1091cad2ce .